### PR TITLE
Tasks left before Poggit is officially released (finally 1.0, not beta!)

### PR DESCRIPTION
**Pull request by SOF3**
- [x] Virion infection while building
- [x] Voting to approve plugins
- [x] Better release list (sorting!) (bug: #126)
  - [ ] Tune the UI: What about making better use of Categories? And more compact buttons?
- [x] User timeline
- [x] Optimization
  - [x] MemberHomePage optimization (#129)

### CODE REVIEW

File: src/poggit/release/review/ReviewManagement.php [hunk 1]
```diff
@@ -40,6 +39,8 @@ protected function impl() {
 
             case "add":
                 $uid = OfficialReviewModule::getUIDFromName($user);
+                if ($_POST["score"] > 5 || $_POST["score"] < 0 || (strlen($_POST["message"] > 256 && $userlevel < Poggit::MODERATOR))) break;
```
Comment (SOF3) [inline src/poggit/release/review/ReviewManagement.php]: Wrong placement of parentheses. Check the arguments of `strlen`.

File: src/poggit/home/MemberHomePage.php [hunk 2]
```diff
@@ -48,17 +48,17 @@ public function __construct() {
...
                         AND builds.class IS NOT NULL ORDER BY created DESC LIMIT 1) AS builddate,
                 (SELECT COUNT(*) FROM builds WHERE builds.projectId=p.projectId 
                         AND builds.class IS NOT NULL) AS bcnt,
                 IFNULL((SELECT CONCAT_WS(',', buildId, internal) FROM builds WHERE builds.projectId = p.projectId
                         AND builds.class = ? ORDER BY created DESC LIMIT 1), 'null') AS bnum
-                FROM projects p INNER JOIN repos r ON p.repoId=r.repoId WHERE r.build=1 ORDER BY r.name, pname", "i", ProjectBuilder::BUILD_CLASS_DEV) as $projRow) {
+                FROM projects p INNER JOIN repos r ON p.repoId=r.repoId WHERE r.build=1 AND $where ORDER BY r.name, pname", "i", ProjectBuilder::BUILD_CLASS_DEV) as $projRow) {
             $repo = isset($repos[(int) $projRow["rid"]]) ? $repos[(int) $projRow["rid"]] : null;
-            if(!isset($repo)) continue;
+            if(is_null($repo) || in_array($repo, $this->repos ?? [])) continue;
```
Comment (SOF3) [inline src/poggit/home/MemberHomePage.php]: What about indexing by repoId?

File: src/poggit/release/review/ReviewManagement.php (hunk 1 above)
Comment (Awzaw) [reply src/poggit/release/review/ReviewManagement.php]: Oops

File: src/poggit/timeline/TimeLineEvent.php [hunk 3]
```diff
@@ -70,6 +70,6 @@ public function dispatchFor(int $uid) {
 
     public function dispatch(): int {
         return MysqlUtils::query("INSERT INTO event_timeline (type, details) VALUES (?, ?)",
-            "is", $this->getType(), json_encode($this))->insert_id;
+            "is", $this->getType(), json_encode((array) $this))->insert_id;
```
Comment (SOF3) [inline src/poggit/timeline/TimeLineEvent.php]: Actually I made the jsonSerialize() function to unset `eventId` and `created`, which are already in the database.

File: doc/mysql.sql [hunk 4]
```diff
@@ -41,7 +41,8 @@ DROP TABLE IF EXISTS project_subs;
 CREATE TABLE project_subs (
     projectId INT UNSIGNED REFERENCES projects(projectId),
     userId INT UNSIGNED REFERENCES users(uid),
-    level TINYINT DEFAULT 1 -- New Build = 1
+    level TINYINT DEFAULT 1, -- New Build = 1
+    UNIQUE KEY user_project (userId, projectId)
```
Comment (SOF3) [inline doc/mysql.sql:52]: 👍 I forgot to check this

File: js/releaseDetails.js [hunk 5]
```diff
@@ -100,7 +100,7 @@ $(function() {
             if(skipGeneral && i === 0) continue;
             titleTabs.append($("<li></li>").append($("<a></a>")
                 .attr("href", "#" + ids[i])
-                .text(titles[i])));
+                .text(titles[i].substr(0, 40) + (titles[i].length > 40 ? "..." : ""))));
```
Comment (SOF3) [inline js/releaseDetails.js]: But I am not moving the header to the tab content, show the `...` will never be shown.

File: doc/mysql.sql [hunk 6]
```diff
@@ -348,7 +348,7 @@ CREATE FUNCTION KeepOnline(p_ip VARCHAR(40), p_uid INT UNSIGNED)
         ON DUPLICATE KEY UPDATE lastOn = CURRENT_TIMESTAMP;
 
         DELETE FROM users_online
-        WHERE UNIX_TIMESTAMP() - UNIX_TIMESTAMP(lastOn) < 300;
+        WHERE UNIX_TIMESTAMP() - UNIX_TIMESTAMP(lastOn) > 300;
```
Comment (SOF3) [inline doc/mysql.sql]: 🙈

File: js/newSubmit.js [hunk 7]
```diff
@@ -1322,24 +1326,22 @@ Do you still want to save this draft?`)) return;
...
                                 var dateSpan = $("<span></span>").attr("data-timestamp", versions[relId].submitTime);
                                 timeTextFunc.call(dateSpan.get()[0]);
                                 // noinspection JSUnresolvedVariable
-                                dialogData.table.append($("<tr></tr>").addClass("submit-deps-version-select-row")
+                                dialogData.table.prepend($("<tr></tr>").addClass("submit-deps-version-select-row")
                                     .append($("<td></td>")
                                         .append($("<input type='radio'/>").addClass("submit-dep-dialog-version-radio").attr("name", versionDialogLock)
                                             .attr("data-value", relId)
-                                            .prop("checked", first))
+                                            .prop("checked", true))
```
Comment (SOF3) [inline js/newSubmit.js:1337]: Won't this check every radio button?

Comment (Awzaw) [reply js/newSubmit.js:1337]: Not if they are grouped as only one per radio group can be checked - this leaves only the latest selected

**Comment by VCraftMCPE** [comment]
what about a new design

**Comment by SOF3** [comment]
Could you give some practical suggestions?

**Comment by pranavbhuv** [comment]
https://www.google.co.uk/search?safe=strict&biw=1160&bih=934&tbm=isch&sa=1&ei=F2JWWtTWJMrGgAbutLGoAg&q=ui&oq=ui&gs_l=psy-ab.3..0i67k1l4j0l6.8110.8355.0.8736.2.2.0.0.0.0.91.158.2.2.0....0...1c.1.64.psy-ab..0.2.157....0.vemqTaNEiu8#imgrc=MpFDoF7aF6g5QM:

**Comment by SOF3** [comment]
How is that relevant?

**Comment by dktapps** [comment]
Is this thing serving any purpose at this point? I only mention it because it keeps notifying me... no big deal I can unsubscribe, just seems pointless to keep this open

**Comment by SOF3** [comment]
True, versioning doesn't make sense for Poggit unless we're rewriting it.

**Pull request closed without merging**